### ISSUE TITLE
[css-grid] Decide on a name for `item-slack`

### ISSUE THREAD

**Issue by Loirooriol** [w3c/csswg-drafts#10884]
https://github.com/w3c/csswg-drafts/issues/9328#issuecomment-1823217994 didn't resolve on the name for this property.

In the spec it has been tentatively added as `masonry-slack`:  https://drafts.csswg.org/css-grid-3/#masonry-slack

Other ideas from #9328 are `masonry-threshold`, `masonry-placement-threshold`, `masonry-strictness`, `masonry-adjust`, `masonry-skip`, `masonry-sensitivity`.

**Comment by Loirooriol** [comment]
I'm liking `masonry-tolerance` (as in https://en.wikipedia.org/wiki/Engineering_tolerance).

I'm not fond of "slack" because as a non-native English speaker that's a word that I learned considerably later than the others.

"strictness" and "sensitivity" can seem to imply that a higher value will choose the track by comparing the running positions more strictly, while it's actually the opposite.

**Comment by chriskirknielsen** [comment]
I really like `item-tolerance` [as suggested in the item-flow proposal post](https://webkit.org/blog/16587/item-flow-part-1-a-new-unified-concept-for-layout/#:~:text=item-tolerance), as it maps to a physical, real-world concept (afaik used in manufacturing specs), whereas "slack", feels a little more… "jargony", if that makes sense?

**Comment by astearns** [comment]
I wonder whether more information about what gets the slack would improve the property name. I see that this was suggested in the earlier issue, using `-placement-`, which is a bit long (but I still like it). Perhaps `-fit-` would work?

`item-fit-slack`
`item-fit-tolerance`
`item-fit-allowance`

**Comment by davidwebca** [comment]
There is ```item-fit-clearance```, ```item-fit-buffer```, ```item-fit-room```, but I like ```item-fit-tolerance``` the best because it very much describes what it does without too much potential interference from other meanings (buffer = video buffer? / clearance = clearing css floats?).

Other funny suggestions : ```item-fit-wiggle-room```,  ```item-fit-leeway```, ```item-fit-laxity```.

**Comment by DavidJCobb** [comment]
The name definitely needs to at least imply that the property pertains to how items are fit into the layout -- specifically, the concept of "fitting" a thing in. Things like "placement" don't, and absent any other qualifiers, words like "tolerance" and "threshold" are way too vague. (What is it the threshold *of*?) As far as the two-word suggestions go, `item-fit-tolerance` is the closest to being good.

The word "slack" conveys looseness, whereas it seems like higher values would cause items to be fit into the layout in more tightly compressed arrangements. So "more slack" leads to a tighter, not looser, fit? Seems backwards.

[WebKit's explainer](https://webkit.org/blog/16587/item-flow-part-1-a-new-unified-concept-for-layout/) describes the property as inviting the browser to "cram" extra items into tight spaces. Maybe `item-cram` or `item-cram-tolerance`? The word "cram" isn't super formal, but it well conveys the idea of squeezing things into too tight a space for them to fit comfortably. The word "fit" is less specific about this. I do sort of think that mentioning "tolerance" is good to distinguish this from e.g. a per-item or overall "limit" on how things can be crammed (and it leaves room to actually add a "limit" property in the future if that's ever deemed a good idea).

**Comment by davidwebca** [comment]
I think the focus needs to be on describing the extra space and not the result of that extra space to keep the property semantically distinct.  `item-cram-tolerance` describes only one of the potential result - cramming more elements in a specific space - but according to the explainer, and if I understand correctly, that's not necessarily what would happen in a grid/masonry scenario. The name of the property, in my opinion, needs to say: "Here's a value that describes extra space for your calculation, wether it be for compacting or positioning the next item." When I think about it that way, I'd rather forgo the `fit` as a prefix and it makes me consider `ìtem-fit` alone, `item-boundary`, `item-cap`, `item-edge`, `item-bleed` or `item-safety` (those last two inspired by print design).

Just to sum up my understanding:

In flexbox, a higher number will allow a larger element to be compacted into the current row before the algorithm moves to the next. If in the current row there is `20em` of space left, if the `item-tolerance` is `10em` and if the next item to place is `30em`, that item will be crammed into the `20em` that's left with potential recalculations depending on the packing, spacing and stretching properties. With an `item-tolerance` of `9em`, the cutoff point would be at `29em` and so the next item would be placed on a new row.

In grid, a higher number will allow to re-equalize columns in a specific direction. If column 1 has `10em` of free space, column 2 has `20em` of free space, the next item should usually be placed in column 2 to fill the column that has the most empty space. An `item-tolerance` of `10em` would place the next item in column 1 instead of column 2 because both columns would now be considered having `20em` free space - it effectively "virtually" backfills the empty space with the value we provide and then restarts its column choice algorithm.

 Please correct me if I'm wrong in my current understanding.

But if all of this is right, `item-tolerance`, `item-fit` alone and `item-threshold` are much more specific in what they accomplish while being generic enough to accommodate both display-mode paradigms graciously.

**Comment by fantasai** [comment]
Edited in the renaming to item-tolerance, but leaving this issue open to collect other ideas given the minutes indicating openness to better names.

**Comment by fantasai** [comment]
If we're not going with the `item-flow` set of flow-control properties, `item-tolerance` is not a very good name. (Even then, it was a bit awkward.) It's not about tolerance on the item (like its size or whatever), but about tolerance in placement. So I think we need a better name. Some ideas to get started:

- `flow-tolerance`
- `fit-tolerance`
- `pack-tolerance`
- `placement-tolerance`

**Comment by markdavidmc0** [comment]
My 2c is `order-tolerance` because the lane ordering is decided by the tolerance.

**Comment by Loirooriol** [comment]
Is this going to be specific to grid-lanes layout? If so, `grid-lanes-tolerance`?

I recall some ideas to make it work in flex layout, but I'm not sure if they went anywhere.

**Comment by nathanchase** [comment]
What about `flow-margin`?

To use @jensimmons's car metaphor - imagine every car has an invisible, adjustable rear bumper extension - the **_flow-margin_**.

Low flow-margin: The bumpers are tight. If Car 4 is even an inch shorter than Car 1, the driver of Car 6 sees that "gap" and swerves into the right lane to get ahead.

High flow-margin: The bumpers are long. Even if Car 4 is shorter, its flow-margin covers the difference. Car 6 stays put because, according to the "margin," there’s no meaningful room to move forward.

**Comment by Loirooriol** [comment]
@nathanchase  If you don't link to that car metaphor, out of context your comment isn't making much sense to me.

Anyways this is definitely not the same as adding a margin to every item, so I don't like it.

**Comment by nathanchase** [comment]
> Anyways this is definitely not the same as adding a margin to every item, so I don't like it.

I mean, it's about how much room the item has or doesn't have to take up the next available position, no?

<img width="1536" height="864" alt="Image" src="https://github.com/user-attachments/assets/ccb09eb9-f90b-41ab-bf73-174267ef0270" />

It seems to me - at least based on the demos thus far - that more often than not if people employ this "tolerance", it will be to achieve a new order of items within the lanes because the default order seems "wrong".

So the value is some measure of this "wrongness"... tolerance, margin of error, etc. 

It seems sensible to think that the lanes have either more or less space for items to occupy based on this value.

**Comment by dennisfrank** [comment]
How about `item-safe-space`? 

That’s just my first idea, off the top of my head. I think it describes the behaviour quite well without being too technical/complicated.

**Comment by philiprenich** [comment]
tl;dr: `flow-tolerance`

I think "item" can be confusing. I usually need to look up to remind myself the difference between `align-/justify-` `content/items`

And in those cases, "items" is setting the default "self" property, so it's actually more related to the child than the parent. In our case here for grid-lanes, is there going to be a correlated `tolerance-self`? If not, I would say stay away from "item" as that causes a usage overload.

`tolerance` is a good choice. `margin` is already in usage and means something else (ie, this property won't change the distance between items like `margin` does). It's also a single word which I like.

I think of the ideas I've seen floated in the last couple days of posts here (when I joined in) is `flow-tolerance`. `fit` isn't bad, but we're deciding how items will be placed in the flow, not how close. `pack` is just used in the same way as `fit` imo.

**Comment by CarolMcKay** [comment]
layout-tolerance feels like a good name and flexible enough for future layout models.

**Comment by acobster** [comment]
I like `flow-tolerance` as well, or `masonry-tolerance`.

`order-tolerance` and `pack-tolerance` are both okay, but don't describe what's actually happening as precisely IMHO. Flow in this context already means the flow of content along the width/height of the document, modulo any obstacles. The obstacles in this case just happen to be slightly longer items. So "flow tolerance" to me suggests a maximum tolerance for deviating from what the flow algorithm would strictly do otherwise, which is exactly what we want. I think `masonry-tolerance` works for the same reason, but isn't quite as obvious.

"Order" suggests something similar but could also suggest the opposite: how tolerant to be of placing items _out_ of order due to some hypothetical factor, which is confusing to me. And any item can "fit" in any lane, so I think this doesn't clarify what we're tolerating. "Pack" is a little closer but also seems pretty similar to fit.

I think `flow-margin` muddies the waters. It obscures the tolerance concept, which is really what we're after. And what we're tolerating isn't an item's margin, it's the _adjustment_ to that margin, i.e. how much we can adjust the margin to achieve a smoother flow. And margin is already overloading an existing term, so I find it extra confusing. I would expect something with `margin` in the name to apply consistently to each item, but instead what the property is describing is, in a sense, how inconsistent we can be.

**Comment by jensimmons** [comment]
Hello everyone coming over from the [article about Grid Lanes](https://webkit.org/blog/17660/introducing-css-grid-lanes/)! 👋

It‘s likely that this property for tolerance will find its way into Flexbox as well. And work for the original CSS Grid. Perhaps for other layout modes as well… so the name should be flexible enough for such a future. 

Also, if you scroll to the top of this issue, you can see that the CSS Working Group already considered a lot of options for the second half of this property and resolved on `*-tolerance`. Mostly the question now is what should the first half of the name be? It was `item-` because it was going to be part of a whole group of properties that start with `item` — but that's not happening anymore, so it makes sense to reconsider the first word in the name.

**Comment by nathanchase** [comment]
In that case, if verbosity isn't an issue, `placement-tolerance` is very apt and descriptive and would apply across flexbox and grid.

`fit-tolerance` could be confused with `object-fit`, potentially - for someone new to CSS - despite vastly different applications.

`order-tolerance` could be confused with `order`. 
(The fact that `order` is not `grid-order` might be convincing enough precedence to just use `tolerance` without a prefix?)

**Comment by Loirooriol** [comment]
> It‘s likely that this property for tolerance will find its way into Flexbox as well. And work for the original CSS Grid

For flexbox it could e.g. allow placing an item into a line even if it would overflow a bit (with the idea that items will then shrink).

But how would it work in grid? I don't recall any idea in this direction.

**Comment by nt1m** [comment]
WPT changes are here: https://github.com/web-platform-tests/wpt/pull/57063

**Comment by nt1m** [comment]
Spec PR: https://github.com/w3c/csswg-drafts/pull/13313

**Comment by Crissov** [comment]
It seems strange to me that `flow-tolerance` would have nothing to do whatsoever with the similar sounding inner `display` values `flow` and `flow-root`, because those are orthogonal to `grid-lanes` (and `flex`). 

PS I do realize that the parent’s `grid-auto-flow` property and the `auto-flow` keyword in `grid` or the `flex-flow` shorthand property are related and the spec sometimes uses _flow_ regarding auto-placement and _in-flow_ for relevant children of the grid container. There, _flow_ is always appended and thus clarified by context. Making it the first hyphenated part of a property is what trips me off. 
My suggestion would be `track-tolerance` since `lane-tolerance` would be too specific. Alas, Flexbox doesn’t (need to) talk about _tracks_ at all, more or less using _axis_ instead.

**Comment by carnoxen** [comment]
> It seems strange to me that `flow-tolerance` would have nothing to do whatsoever with the similar sounding inner `display` values `flow` and `flow-root`, because those are orthogonal to `grid-lanes` (and `flex`).

Moreover, `flow-*` properties was proposed in [CSS Regions](https://www.w3.org/TR/css-regions-1/#the-flow-into-property) (however it was discontinued).

**Comment by mirisuzanne** [comment]
The meeting notes mention that "Discussion on item-slack in Flexbox is at https://github.com/w3c/csswg-drafts/issues/11480" but that issue was closed as a 'unifying theory' and again tolerance in flexbox is mentioned as a smaller use-case that might deserve more exploration. But I don't think that issue was ever opened. Or has that idea been abandoned at this point?

Edit: this is now in https://github.com/w3c/csswg-drafts/issues/3071

**Comment by alisonmaher** [comment]
I don't want to spend too much time relitigating this issue, but given some of the author confusion around this property, re-opening to discuss the following options and decide if we want to keep or change the name (hopefully one last time!):

1. Keep `flow-tolerance` - this already ships in Webkit, but authors have found this property confusing, and it could be confused with inner `display` values `flow` and `flow-root` (and maybe `flow-*` properties in CSS Regions which is likely less of a concern).
2. `fit-tolerance` - this naming seems like it may translate nicely to the [flex equivalent use case](https://github.com/w3c/csswg-drafts/issues/3071). Previous pushback on "fit" was confusion with `object-fit`, but that seems less concerning than some of the push back raised around the use of `flow-*`.
3. `fill-tolerance` would explain the concept of what this controls in grid lanes more clearly, but it could be confused with the existing `fill` property.

Given the various options and tradeoffs, `fit-tolerance` seems like it might have the least confusion with other concepts in CSS, while more clearly explaining what the property is impacting. However, as noted, this would mean Webkit would need to consider aliasing to the new naming.

CC @fantasai

**Comment by chriskirknielsen** [comment]
With CSS grid, you can use `repeat(auto-fit|auto-fill, ...)`, so would that also cause confusion? And agreed on `fill` being more confusing, I associate that with colour/SVG, and it'd definitely look weird alongside `fill` and friends.

If `flow-tolerance` is off the table, would `place-tolerance` work as a less confusing choice? (with the slight nitpick that `place-*` are shorthands to set both directions, and this `*-tolerance` only affects one axis…right?)

**Comment by alisonmaher** [comment]
> With CSS grid, you can use repeat(auto-fit|auto-fill, ...), so would that also cause confusion?

Good point, another reason why "fill" here could be confusing.

> would place-tolerance work as a less confusing choice?

`placement-tolerance` was an earlier suggestion, but I think it was considered too verbose. "place" would be shorter.

> with the slight nitpick that place-* are shorthands to set both directions, and this *-tolerance only affects on axis…right?

Yeah, at least for the existing layout modes where this would apply, it would only apply to one dimension
